services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -0,0 +1,146 @@
+    /** Hakijan tiedot */
+    val customerWithContacts: HakemusUpdateRequestCustomerWithContacts?
+    /** Asianhoitajan tiedot */
+    val representativeWithContacts: HakemusUpdateRequestCustomerWithContacts?
+}
+
+data class JohtoselvityshakemusUpdateRequest(
+    // 1. sivu Perustiedot (first filled in Create) //
+    /** Nimi */
+    override val name: String,
+    /** Katuosoite */
+    override val streetAddress: String,
+    /** Työssä on kyse: Uuden rakenteen tai johdon rakentamisesta */
+    val constructionWork: Boolean,
+    /** Työssä on kyse: Olemassa olevan rakenteen kunnossapitotyöstä */
+    val maintenanceWork: Boolean,
+    /** Työssä on kyse: Kiinteistöliittymien rakentamisesta */
+    val propertyConnectivity: Boolean,
+    /**
+     * Työssä on kyse: Kaivutyö on aloitettu ennen johtoselvityksen tilaamista merkittävien
+     * vahinkojen välttämiseksi
+     */
+    val emergencyWork: Boolean,
+    /** Louhitaanko työn yhteydessä, esimerkiksi kallioperää? */
+    val rockExcavation: Boolean,
+    /** Työn kuvaus */
+    val workDescription: String,
+    // 2. sivu Alueet
+    /** Työn arvioitu alkupäivä */
+    override val startTime: ZonedDateTime? = null,
+    /** Työn arvioitu loppupäivä */
+    override val endTime: ZonedDateTime? = null,
+    /** Työalueet */
+    override val areas: List<ApplicationArea>? = null,
+    // 3. sivu Yhteystiedot
+    /** Hakijan tiedot */
+    override val customerWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    /** Työn suorittajan tiedot */
+    val contractorWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    /** Rakennuttajan tiedot */
+    val propertyDeveloperWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    /** Asianhoitajan tiedot */
+    override val representativeWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    // 4. sivu Liitteet (separete endpoint)
+    // 5. sivu Yhteenveto (no input data)
+) : HakemusUpdateRequest {
+
+    /**
+     * Returns true if this application data has changes compared to the given [applicationData].
+     */
+    fun hasChanges(applicationData: CableReportApplicationData): Boolean =
+        name != applicationData.name ||
+            streetAddress != applicationData.postalAddress?.streetAddress?.streetName ||
+            constructionWork != applicationData.constructionWork ||
+            maintenanceWork != applicationData.maintenanceWork ||
+            propertyConnectivity != applicationData.propertyConnectivity ||
+            emergencyWork != applicationData.emergencyWork ||
+            rockExcavation != applicationData.rockExcavation ||
+            workDescription != applicationData.workDescription ||
+            startTime != applicationData.startTime ||
+            endTime != applicationData.endTime ||
+            areas != applicationData.areas ||
+            customerWithContacts?.hasChanges(applicationData.customerWithContacts)
```

Comment:
With kortisto, `CableReportApplicationData` will not have `customerWithContacts`, since the customers and contacts will be saved separately to DB.

I think the comparison should be done against ApplicationEntity.

Reply:
Changing the name fields is nice if we also change the frontend to match.

Reply:
Since the frontend needs to be changed for contacts and contact persons I think these other changes would go with the same effort. The data structures that UI uses internally may still remain as they are, only creating the update request would need updating.

Reply:
That is a good and valid point. Would it be too much changes to both backend and frontend if the new HakemusController endpoints would return a flatter structure with Finnish naming instead of current Application+ApplicationData structure with English/Allu naming? Maybe refactoring to flat+Finnish is something for the future and I should revert my changes in this update request also to follow ApplicationData+English naming?

Reply:
I will revert these flat+Finnish changes at this point